Shell add getopt_long

jakub-uC opened the pull request:
This PR includes the following changes:
1. Remove `shell_getopt` and replace it with simple `getopt` to be compatible with `getopt` usage.
2. Add support for `getopt_long` and `getopt_long_only`
3. Add tests for getopt library functions

nordic-krch commented on include/shell/shell.h:
```diff
@@ -1010,6 +1011,56 @@ int shell_getopt(const struct shell *shell, int argc, char *const argv[],
  */
 struct getopt_state *shell_getopt_state_get(const struct shell *shell);
+
+#if defined CONFIG_SHELL_GETOPT_LONG
```
API shall not be ifdefed: https://docs.zephyrproject.org/latest/guides/design_guidelines.html#conditional-data-and-apis
`Function declarations that are available only when the option is enabled should be provided unconditionally`

nordic-krch commented on include/shell/shell.h:
```diff
@@ -1010,6 +1011,56 @@ int shell_getopt(const struct shell *shell, int argc, char *const argv[],
+ *			   @ref struct option.
+ * @param[in] long_idx	   If long_idx is not NULL, it points to a variable
+ *			   which is set to the index of the long option relative
+ *			   to @ref long_options.
```
probably should be `@p long_options`

nordic-krch replied:
still `@ref`

jakub-uC replied:
nope, it is as it is :)

nordic-krch commented on include/shell/shell.h:
```diff
@@ -1010,6 +1011,56 @@ int shell_getopt(const struct shell *shell, int argc, char *const argv[],
+ * @note This function does not support environment variable: POSIXLY_CORRECT.
+ *
+ * It is based on FreeBSD implementation.
+ *
```
arguments documentation missing

nordic-krch replied:
`@ref long_options` -> `@p long_options`

jakub-uC replied:
done

nordic-krch commented on lib/util/getopt/getopt.h:
```diff
@@ -33,6 +38,44 @@ void getopt_init(struct getopt_state *state);
 int getopt(struct getopt_state *const state, int nargc,
 	   char *const nargv[], const char *ostr);
 
+#if CONFIG_GETOPT_LONG
```
remove ifdef from header.

nordic-krch reviewed (commented):
Could you add any test or sample which is using this feature?

jakub-uC commented:
@carlescufi @nordic-krch 
In order to be able to perform any tests, we need to change the name of the library function `getopt` to `z_getopt` to avoid naming conflicts. There is already an implementation of `getopt` in GNU.

nordic-krch commented on lib/util/getopt/getopt.c:
```diff
@@ -39,7 +39,93 @@ LOG_MODULE_REGISTER(getopt);
+	int ret;
+
+	state_store(&state);
+	ret = z_getopt(&state, nargc, nargv, ostr);
```
why not using global `struct z_getopt_state` all the time without need for store and restore?

jakub-uC replied:
Unfortunately we can't do it this way because the normal use of the getopt function requires global variables.

carlescufi commented:
@jakub-uC could you please fill out an issue as described [here](https://docs.zephyrproject.org/latest/contribute/external.html#submission-and-review-process)?

Commit:
lib: getopt: rework and extend getopt library

Getopt has been rework in this way that calling it does not require
extra state parameter and its execution is thread safe.
Global parameters describing the state of the getopt function have been
made available to ensure full API compatibility in using this library.
However, referencing these global variables directly is not thread
safe. In order to get the state of the getopt function for the thread
that is currently using it, call: getopt_state_get();

Extended the library with getopt_long and getopt_long_only functions.

Moved getopt libary from utils to posix.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
doc: shell getopt support

Documentations has been updated acorcing to changes in the getopt
library.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

Commit:
tests: lib: getopt

Add tests for the getopt library.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

cfriedt commented:
My apologies for just adding my 2c now but I hope I can ask a potentially dumb question.

Why not put this implementation of `getopt()` into `lib/libc/minimal/source/stdlib`? Newlib also provides `getopt()`, `getopt_long()`, and `getopt_long_only()` and keeps it inside of the `stdlib` directory.

https://github.com/bminor/newlib/search?q=getopt

Am I missing some other important piece of information?

jakub-uC commented:
@cfriedt Thank you for the input. When I've started working on this PR was looking for the `getopt` implementation in the past and there were 2 proposals:
1. FreeBSD -> main and first proposal. In fact existing getopt is based on it.
2. NuttX -> it was proposed during getopt_long and getopt_long only implementation because it has license compliant with Zephyr.
I've spent plenty of time integrating both with Zephyr. My conclusion is that `getopt_long` and `getopt_long_only` from NuttX are not performing very well. Source code from NuttX is not passing the unit tests based on https://www.gnu.org/software/libc/manual/html_node/Getopt-Long-Option-Example.html while FreeBSD is. 
Hence I've focused on FreeBSD.

This was my side task and in the next weeks I will not be able to work on your proposal. Nevertheless, would you be so kind and let me know if the license of https://github.com/bminor/newlib/search?q=getopt is compliant with Zephyr?

cfriedt commented:
@jakub-uC - I'm not implying we copy the newlib version into Zephyr, because we probably link against it anyway when `CONFIG_NEWLIB_LIBC` is selected. I'm only suggesting that maybe placement of your implementation might be best to go into the minimal libc rather than lib/util, because getopt normally is part of the libc

carlescufi commented:
@cfriedt getopt is not part of the C standard library, but rather part of POSIX, hence its placement outside `lib/libc/minimal`.

carlescufi commented:
@jakub-uC could you please fix the Coding Guidelines and Compliance issues?

carlescufi commented:
Note that this is an extension of the original `getopt` inclusion in Zephyr, which was already approved by the TSC and got no objections from the board.

cfriedt commented:
Previously this implementation was called shell_getopt(), but now it's overriding the standard (POSIX) version.

I think that's fine when the minimal libc is selected (and location... sure), but is there some technical reason against using the newlib version of these functions when we link with newlib?

carlescufi commented:
> Previously this implementation was called shell_getopt(), but now it's overriding the standard (POSIX) version.

That is true.

> I think that's fine when the minimal libc is selected (and location... sure), but is there some technical reason against using the newlib version of these functions when we link with newlib?

I think there is, the fact that we do not implement the standard stdin/stdout. Maybe we should rename to `shell_getopt`? @jakub-uC ?

cfriedt commented:
> > Previously this implementation was called shell_getopt(), but now it's overriding the standard (POSIX) version.
> 
> That is true.
> 
> > I think that's fine when the minimal libc is selected (and location... sure), but is there some technical reason against using the newlib version of these functions when we link with newlib?
> 
> I think there is, the fact that we do not implement the standard stdin/stdout. Maybe we should rename to `shell_getopt`? @jakub-uC ?

@carlescufi - getopt does not use stdin / stdout.

carlescufi commented:
> > > Previously this implementation was called shell_getopt(), but now it's overriding the standard (POSIX) version.
> > 
> > 
> > That is true.
> > > I think that's fine when the minimal libc is selected (and location... sure), but is there some technical reason against using the newlib version of these functions when we link with newlib?
> > 
> > 
> > I think there is, the fact that we do not implement the standard stdin/stdout. Maybe we should rename to `shell_getopt`? @jakub-uC ?
> 
> @carlescufi - `getopt()` does not use stdin / stdout. It's really just operating on strings (but it does make use of global variables, which does mean that there are side-effects).
> 
> One slight gotcha is that `argv[0]` might not refer to anything useful in our implementation (normally it refers to the path of the currently executing program). That should be more or less expected though, based on the [man 3 getopt](https://man7.org/linux/man-pages/man3/getopt.3.html).

Good point @cfriedt. In that case I suggest we keep the `getopt` name when using minimal libc, and make the Kconfig option automatically deselect it when using newlib.

cc @stephanosio

stephanosio reviewed (commented):
Some general questions and topics for discussion:
1. `getopt` in itself is a function defined in the POSIX standard. Noting that, should this implementation be part of the Zephyr POSIX library (`lib/posix/`)?
2. Newlib has its own `getopt` and `getopt_long` implementation. Assuming this feature will depend on `!NEWLIB_LIBC` due to conflicts, would the `SHELL_GETOPT` seamlessly switch to using the newlib implementation when newlib is enabled? If not, are newlib users doomed to never use `SHELL_GETOPT`?
3. In general, are we heading in the right direction implementing these functions on the Zephyr-side? `getopt` is not really something for highly resource-constrained devices and its intended use would be in "moderately resourceful devices" that can accommodate the newlib. Assuming newlib can provide these, why not just use the newlib implementation and make `NEWLIB_LIBC` a prerequisite for `SHELL_GETOPT`? Is it really worth the effort maintaining these two independent implementations?

cfriedt commented:
> Some general questions and topics for discussion:
> 
> 1. `getopt` in itself is a function defined in the POSIX standard. Noting that, should this implementation be part of the Zephyr POSIX library (`lib/posix/`)?

+1 for that

> 2. Newlib has its own `getopt` and `getopt_long` implementation. Assuming this feature will depend on `!NEWLIB_LIBC` due to conflicts, would the `SHELL_GETOPT` seamlessly switch to using the newlib implementation when newlib is enabled? If not, are newlib users doomed to never use `SHELL_GETOPT`?

I think that applications requiring `getopt(3)` should require one of
* `CONFIG_BOARD_NATIVE_POSIX` or `CONFIG_BOARD_NATIVE_POSIX_64BIT` or `CONFIG_BOARD_NRF52_BSIM`
* `CONFIG_POSIX_API`
* `CONFIG_NEWLIB_LIBC`

Right now, I believe the only consumer of `getopt(3)` is the shell, so the addition of this in Kconfig is pretty trivial.

@jakub-uC - I appreciate that you put a lot of work into this, and I would like to see it get merged. Please move those few files around and make the necessary changes.

cfriedt reviewed (changes requested):
I would be happy to approve / merge this because I think it's fairly useful, but only on the condition the implementation is moved to lib/posix

jakub-uC commented:
@cfriedt : done

Commit:
CODEOWNERS: update getopt library location

Getopt has been moved from /lib/utils to /lib/posix hence
CODEOWNERS must be updated accordingly.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

cfriedt commented:
@jakub-uC - a few build errors that need to be resolved, otherwise, thank you!

jakub-uC commented:
@carlescufi @cfriedt 
I have a problem with this PR. I have no idea how to build tests for the native_posix boards to use the getop from
posix/getopt/getopt.h instead of newlib/include/getopt.h.

carlescufi commented:
@jakub-uC could you please rebase?

carlescufi commented:
The current issue is that native_posix picks up the system-wide `getopt.h`, which does not include `getopt_init()`, making the build fails. I'd be very grateful if @stephanosio and @cfriedt could weigh in here, has this problem of extending posix headers been solved before in Zephyr?

cfriedt commented:
> The current issue is that native_posix picks up the system-wide `getopt.h`, which does not include `getopt_init()`, making the build fails. I'd be very grateful if @stephanosio and @cfriedt could weigh in here, has this problem of extending posix headers been solved before in Zephyr?

Easiest way to make the build pass is to disable  the tests on `CONFIG_ARCH_POSIX`. We *do* want to use the system `getopt()` on those platforms, and we are not obligated to test that. 

My next suggestion would be to remove `getopt_init()` and `getopt_state`. It is not guaranteed to be reentrant by the spec.

jakub-uC commented:
@carlescufi @cfriedt 
Done.
